### PR TITLE
Working HID driver for Wii U [TESTING NEEDED]

### PR DESCRIPTION
**Pull request by gblues**
## Description

**TESTING NEEDED:** I've tested the digital inputs, but I don't have any cores that use analog inputs handy to test those.

This is a pretty big change to how HID is handled in RetroArch. 

Some background:

RA was designed around the assumption that gamepads and HID devices enjoyed a 1:1 relationship. However, this is not the case. Past HID solutions had implemented special-case handling for specific devices, but I wanted something more generic.

So I made it.

Now, in addition to the joypad driver, there's a HID device driver. The HID device driver can then manage its pad or pads.

These interfaces are implemented as structs of function-pointers, a pseudo-OO approach.

The HID device drivers are designed to be as platform-independent as possible. There's still a small amount of special-casing (like preambles in the activation packets needed for specific platforms), but it's pretty manageable.

The poster child use case for this approach is the GameCube adapter, which exposes 4 controllers through a single HID device. And that's why the Wii U is the lead platform for this new approach.

I was also able to provide some quality-of-life enhancements for WiiU, by making HID pads use the first available port instead of being hard-coded to 5+. Pads will now be activated in the order they connect. The GamePad will still be pad 0, although you can reassign via the menus.

Some of the code is going to look similar to what's already in the `input/connect/` hierarchy. There's some duplication; but since the new drivers are only used on Wii U, it's premature to delete those drivers. Eventually we'll want to port them over to the new style, although it'd be helpful to actually have one of them to test with.

## Reviewers
@aliaspider @QuarkTheAwesome 

[If possible @mention all the people that should review your pull request]

### CODE REVIEW

File: wiiu/input/wiiu_hid.c [hunk 1]
```diff
@@ -613,6 +696,17 @@ static void delete_hidclient(HIDClient *client)
       free(client);
 }
 
+static void init_cachealigned_buffer(int32_t min_size, uint8_t **out_buf_ptr, int32_t *actual_size)
+{
+   int cacheblocks = (min_size < 32) ? 1 : min_size / 32;
+   if(min_size > 32 && min_size % 32 != 0)
+      cacheblocks++;
+
+   *actual_size = 32 * cacheblocks;
```
Comment (aliaspider) [inline wiiu/input/wiiu_hid.c]: `*actual_size = (*actual_size + 0x1F) & ~ 0x1F;` will align to 32bytes.
actual alignment requirement is 64 though, so use `0x3F` instead.

**Comment by twinaphex** [comment]
Works fine for me on OSX - any more additional testing needed?

**Comment by gblues** [comment]
I need to get new builds in the hands of folks at GBATemp so I can be sure I've addressed the problems they reported in the first build. So, not ready to merge yet (but a lot closer!)

**Comment by twinaphex** [comment]
OK, @gblues told me this is fine for merging now so nightly users can get to grips with it, so here goes.

**Pull request merged**